File: write_from_template.py
from pathlib import Path
from string import Template

def write_from_template(file_path, template_name, substitutions, config):
    if not check_file_permission(file_path, "write", config):
        raise PermissionError(f"Write access denied for {file_path}")
    templates = {"basic": "Hello, $name!\nThis is a $type file."}
    template = Template(templates.get(template_name, ""))
    content = template.substitute(substitutions)
    file_path = Path(file_path).resolve()
    with open(file_path, "w", encoding="utf-8") as f:
        f.write(content)
    return f"Wrote template to {file_path}"

File: packages/core/src/tools/write-from-template.ts
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import { promises as fs } from 'fs';
import { resolve } from 'path';
import { checkFilePermission } from '../services/filePermissionService.js';

interface TemplateConfig {
  [key: string]: string;
}

const templates: TemplateConfig = {
  basic: 'Hello, ${name}!\nThis is a ${type} file.',
};

export async function writeFromTemplate(
  filePath: string,
  templateName: string,
  substitutions: { [key: string]: string },
  config: unknown,
): Promise<string> {
  if (!(await checkFilePermission())) {
    throw new Error(`Write access denied for ${filePath}`);
  }

  const template = templates[templateName] || '';
  const content = Object.entries(substitutions).reduce(
    (acc, [key, value]) => acc.replace(`\${${key}}`, value),
    template,
  );

  const resolvedPath = resolve(filePath);
  await fs.writeFile(resolvedPath, content, 'utf-8');
  return `Wrote template to ${resolvedPath}`;
}
